File: unicorn-tp-tools/src/main/java/com/lwohvye/tools/service/impl/EmailServiceImpl.java
/*
 *  Copyright 2019-2020 Zheng Jie
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.lwohvye.tools.service.impl;

import cn.hutool.extra.mail.MailAccount;
import cn.hutool.extra.mail.MailUtil;
import com.lwohvye.core.exception.BadRequestException;
import com.lwohvye.tools.domain.EmailConfig;
import com.lwohvye.tools.domain.vo.EmailVo;
import com.lwohvye.tools.repository.EmailRepository;
import com.lwohvye.tools.service.IEmailService;
import com.lwohvye.core.utils.EncryptUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheConfig;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * @author Zheng Jie
 * @date 2018-12-26
 */
@Service
@RequiredArgsConstructor
@CacheConfig(cacheNames = "email")
public class EmailServiceImpl implements IEmailService {

    private final EmailRepository emailRepository;

    @Override
    @CachePut(key = "'config'")
    @Transactional(rollbackFor = Exception.class)
    public EmailConfig config(EmailConfig emailConfig, EmailConfig old) throws Exception {
        emailConfig.setId(1L);
        if (!emailConfig.getPass().equals(old.getPass())) {
            // 对称加密
            emailConfig.setPass(EncryptUtils.aesEncrypt(emailConfig.getPass()));
        }
        return emailRepository.save(emailConfig);
    }

    @Override
    @Transactional(rollbackFor = Exception.class, readOnly = true)
    @Cacheable(key = "'config'")
    public EmailConfig find() {
        Optional<EmailConfig> emailConfig = emailRepository.findById(1L);
        return emailConfig.orElseGet(EmailConfig::new);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void send(EmailVo emailVo, EmailConfig emailConfig) {
        if (emailConfig.getId() == null) {
            throw new BadRequestException("请先配置，再操作");
        }
        // 封装
        MailAccount account = new MailAccount();
        // 设置用户
        String user = emailConfig.getFromUser().split("@")[0];
        account.setUser(user);
        account.setHost(emailConfig.getHost());
        account.setPort(Integer.parseInt(emailConfig.getPort()));
        account.setAuth(true);
        try {
            // 对称解密
            account.setPass(EncryptUtils.aesDecrypt(emailConfig.getPass()));
        } catch (Exception e) {
            throw new BadRequestException(e.getMessage());
        }
//        部分邮箱 user和fromUser需一致
        account.setFrom(emailConfig.getUser() + " <" + emailConfig.getFromUser() + ">");
        // ssl方式发送-----不使用
//        account.setSslEnable(true);
        account.setSslEnable(false);
        // 使用STARTTLS安全连接
        account.setStarttlsEnable(true);
        String content = emailVo.getContent();
        // 发送
        try {
//            int size = emailVo.getTos().size();
//            Mail.create(account)
//                    .setTos(emailVo.getTos().toArray(new String[size]))
//                    .setTitle(emailVo.getSubject())
//                    .setContent(content)
//                    .setHtml(true)
//                    //关闭session
//                    .setUseGlobalSession(false)
//                    .send();
            MailUtil.send(account, emailVo.getTos(), emailVo.getSubject(), content, true);
        } catch (Exception e) {
            throw new BadRequestException(e.getMessage());
        }
    }

    @Override
    public void send(List<String> tos, String subject, String content) {
        var emailVo = new EmailVo().setTos(tos).setSubject(subject).setContent(content);
        send(emailVo, find());
    }
}

File: unicorn-tp-tools-kotlin/src/main/kotlin/com/lwohvye/tools/service/impl/EmailServiceImpl.kt
/*
 *  Copyright 2019-2020 Zheng Jie
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.lwohvye.tools.service.impl

import cn.hutool.extra.mail.MailAccount
import cn.hutool.extra.mail.MailUtil
import com.lwohvye.core.exception.BadRequestException
import com.lwohvye.core.utils.EncryptUtils
import com.lwohvye.tools.domain.EmailConfig
import com.lwohvye.tools.domain.vo.EmailVo
import com.lwohvye.tools.repository.EmailRepository
import com.lwohvye.tools.service.IEmailService
import org.springframework.cache.annotation.CacheConfig
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * @author Zheng Jie
 * @date 2018-12-26
 */
@Service
@CacheConfig(cacheNames = ["email"])
class EmailServiceImpl(val emailRepository: EmailRepository) : IEmailService {

    @CachePut(key = "'config'")
    @Transactional(rollbackFor = [Exception::class])
    @Throws(Exception::class)
    override fun config(emailConfig: EmailConfig, old: EmailConfig?): EmailConfig {
        emailConfig.id = 1L
        if (emailConfig.pass != old?.pass) {
            // 对称加密
            emailConfig.pass = EncryptUtils.aesEncrypt(emailConfig.pass)
        }
        return emailRepository.save(emailConfig)
    }

    @Transactional(rollbackFor = [Exception::class], readOnly = true)
    @Cacheable(key = "'config'")
    override fun find(): EmailConfig? {
        val emailConfig = emailRepository.findById(1L)
        return emailConfig.orElseGet { EmailConfig() }
    }

    @Transactional(rollbackFor = [Exception::class])
    override fun send(emailVo: EmailVo?, emailConfig: EmailConfig?) {
        if (emailConfig?.id == null) {
            throw BadRequestException("请先配置，再操作")
        }
        // 封装
        val account = MailAccount()
        // 设置用户
        val user = emailConfig.fromUser?.split("@".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray()?.get(0)
        account.user = user
        account.host = emailConfig.host
        account.port = emailConfig.port?.toInt()
        account.isAuth = true
        // 对称解密
        account.pass = EncryptUtils.aesDecrypt(emailConfig.pass)
        //        部分邮箱 user和fromUser需一致
        account.from = " ${emailConfig.user} <${emailConfig.fromUser}>"
        // ssl方式发送-----不使用
//        account.setSslEnable(true);
        account.isSslEnable = false
        // 使用STARTTLS安全连接
        account.isStarttlsEnable = true
        val content = emailVo?.content
        // 发送
        MailUtil.send(account, emailVo?.tos, emailVo?.subject, content, true)
    }

    @Transactional(rollbackFor = [Exception::class], readOnly = true)
    override fun send(tos: List<String>?, subject: String?, content: String?) {
        val emailVo = EmailVo(tos, subject, content)
        send(emailVo, find())
    }
}
